Added Apple Pay, Google Pay and Alternative Payment Methods support

moksonyb opened the pull request:
**New features:**
- Apple Pay support
- Google Pay support
- iDEAL and other APMs support (These are merchant country-dependent. [Learn more](https://developer.sumup.com/online-payments/apm/introduction)
- Some logic to turn off and on APMs and Google Pay
- For Google Pay the integration gets the merchant's business name automatically from SumUp in order to stay consistent

**Documentation:**
- Added a setup guide to both the readme and payment settings page

**Translation**
- Added German translations for the new UI elements

moksonyb commented:
![IMG_2772](https://github.com/user-attachments/assets/fc1ef21f-671d-4862-b9b2-f2f9aa8ca5bb)

wiomoc commented on README.rst:
```diff
@@ -1,9 +1,57 @@
+2. **Apple Pay Setup**:
+
+   * Follow setup steps at `SumUp Wallets Settings <https://developer.sumup.com/settings/wallets>`_ to activate your domain for Apple Pay
+   * Add Apple Developer MerchantID Domain Association file to Pretix Global settings
```
Could you add some description / an example how to do this

moksonyb replied:
I added more detailed instructions to both the settings page and the readme

wiomoc commented on README.rst:
```diff
@@ -1,9 +1,57 @@
+
+   * You must enable Alternative Payment Methods first
+   * Validate your domain with Google (`Learn more <https://developer.sumup.com/online-payments/apm/google-pay>`_)
+   * For the validation screenshots, add ``#sumup-widget:google-pay-demo-mode`` to your payment URL to generate a test Google Pay button
```
I tried to get approval from Google Pay support as well, while experimenting with APMs. I had to go three rounds until I had the screenshot of the checkout accepted. Could you therefore document how you submit the requested screenshots for "Pre-purchase screen", "Payment Method Screen", "Google Pay API payment screen", so that users are spared the back and forth with Google?

moksonyb replied:
Yep, I was confused too first. In the end I ended up submitting the screenshots that I included with one of the last commits. They worked out and I got approved after 24h so I think people can follow them as references.

wiomoc commented on pretix_sumup/payment.py:
```diff
@@ -57,6 +57,85 @@ def settings_form_fields(self):
+                        required=False,
+                        help_text=_(
+                            "<i><strong> You need to enable Alternative Payment Methods to use Google Pay.</strong></i> <br><br>"
+                            if not self.settings.get(
```
Additionally add a `data-checkbox-dependency` like seen in https://github.com/pretix/pretix/blob/e1821f1bb71a32a86f5151f164adf8bce4692698/src/pretix/plugins/paypal2/payment.py#L176 for improved UX

moksonyb replied:
Awesome! I didn't know that you can add dependencies like this :DD
I rewrote the part, now it's much more user friendly

wiomoc commented on pretix_sumup/payment.py:
```diff
@@ -57,6 +57,85 @@ def settings_form_fields(self):
+                        ),
+                        min_length=12,
+                        max_length=18,
+                        disabled=not (
```
UX could be improved with `data-checkbox-dependency` or something similar provided in https://github.com/pretix/pretix/blob/e1821f1bb71a32a86f5151f164adf8bce4692698/src/pretix/static/pretixcontrol/js/ui/main.js#L325-L409

moksonyb replied:
I ended up just using `data-checkbox-dependency` on the APMs checkbox. It's not the perfect solution since an `AND` between the two of them would be better. (Also the dependencies don't propagate in chain so unchecking the `APMs` checkbox will disable the `Enable Google Pay` one but not the `Google Pay Merchant` one if that's dependent on the `Enable Google Pay`)

It can be rewritten in the future but adding the extra custom js script is just not worth it for now.

wiomoc replied:
I raised a PR for supporting transitive dependencies on upstream Pretix - https://github.com/pretix/pretix/pull/5295

wiomoc commented on pretix_sumup/payment.py:
```diff
@@ -68,10 +147,32 @@ def settings_form_clean(self, cleaned_data: dict):
+            merchant_name = cleaned_data.get("payment_sumup_merchant_name")
+
+            if not merchant_id:
+                raise forms.ValidationError(
```
UX could be improved with `add_error` maybe. e.g. https://stackoverflow.com/a/47574536/5048815

moksonyb replied:
Good point! I added the errors to the fields as well :DD

wiomoc commented on pretix_sumup/payment.py:
```diff
@@ -68,10 +147,32 @@ def settings_form_clean(self, cleaned_data: dict):
+                raise forms.ValidationError(
+                    _("Google Pay Merchant ID is required when Google Pay is enabled.")
+                )
+            if not merchant_name:
```
Shouldn't the `merchant_name` always be available if a valid sumup token was provided

wiomoc replied:
Only edge case where this won't hold true would be if updating the plugin with an existing configuration?

moksonyb replied:
Yes, since if you have an older installation the **Merchant Name** won't be saved. You'll need to provide the token again to update it.
It can be changed so if you don't have the merchant name it'll use the saved token to try to get it. I might update it but for now it's a niche edge case

wiomoc commented on pretix_sumup/templates/pretix_sumup/payment_widget.html:
```diff
@@ -52,14 +52,16 @@
 ></script>
 <script type="text/javascript" nonce="{{ csp_nonce }}">
-    SumUpCard.mount({
+    var sumupConfig = {
```
nit: `var` -> `const`

moksonyb replied:
Good point, I changed the whole section so it'll just use Django's template tags in-line

wiomoc reviewed (commented):
Looks like you were quicker than me with APM support :) Many thanks for this PR, I still have a few small comments - but otherwise it looks fantastic!

I wonder if https://github.com/wiomoc/pretix-sumup-payment/blob/0fdc24aac9326b58803d9afc4d529a39fb7bf93f/pretix_sumup/payment.py#L207-L210, https://github.com/wiomoc/pretix-sumup-payment/blob/0fdc24aac9326b58803d9afc4d529a39fb7bf93f/pretix_sumup/payment.py#L218-L223 etc. still behave as expected, especially with non credit card based APMs like IDEAL

wiomoc commented:
> ![IMG_2772](https://private-user-images.githubusercontent.com/31571293/461570875-fc1ef21f-671d-4862-b9b2-f2f9aa8ca5bb.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE0OTAwNjAsIm5iZiI6MTc1MTQ4OTc2MCwicGF0aCI6Ii8zMTU3MTI5My80NjE1NzA4NzUtZmMxZWYyMWYtNjcxZC00ODYyLWI5YjItZjJmOWFhOGNhNWJiLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzAyVDIwNTYwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM2ZTIxZDQwYzFjNzYxODA1ZTA4MWNmNTZmMTU0NTE3OTU2NDUwM2UxNzg1NmRjNjk2YTAxNDZmYjViZWFkYTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MSl2A2d4JkwGkUcAGwK5Z6u4ZkKk8PYCFCL8Xv_iiP8)

Feel free to add this or a screenshot where you can see a bit more of the page to the readme

moksonyb commented:
> Looks like you were quicker than me with APM support :) Many thanks for this PR, I still have a few small comments - but otherwise it looks fantastic!
> 
> I wonder if
> 
> https://github.com/wiomoc/pretix-sumup-payment/blob/0fdc24aac9326b58803d9afc4d529a39fb7bf93f/pretix_sumup/payment.py#L207-L210
> 
> ,
> https://github.com/wiomoc/pretix-sumup-payment/blob/0fdc24aac9326b58803d9afc4d529a39fb7bf93f/pretix_sumup/payment.py#L218-L223
> 
> etc. still behave as expected, especially with non credit card based APMs like IDEAL

An excellent catch! 
Turns outSumUp's API won't return a card field if you pay by iDEAL for instance. Using `transaction["card"]["type"]` in this case just crashed the whole page since it was assuming that you had those fields. 
I changed it to `transaction.get()` everywhere and added an extra field for the payment method with some logic so now it displays the `card_type`, `payment_type` and `last_4_digits` if they exists, otherwise it just leaves it blank.

moksonyb commented:
> > ![IMG_2772](https://private-user-images.githubusercontent.com/31571293/461570875-fc1ef21f-671d-4862-b9b2-f2f9aa8ca5bb.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE0OTAwNjAsIm5iZiI6MTc1MTQ4OTc2MCwicGF0aCI6Ii8zMTU3MTI5My80NjE1NzA4NzUtZmMxZWYyMWYtNjcxZC00ODYyLWI5YjItZjJmOWFhOGNhNWJiLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzAyVDIwNTYwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM2ZTIxZDQwYzFjNzYxODA1ZTA4MWNmNTZmMTU0NTE3OTU2NDUwM2UxNzg1NmRjNjk2YTAxNDZmYjViZWFkYTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MSl2A2d4JkwGkUcAGwK5Z6u4ZkKk8PYCFCL8Xv_iiP8)
> 
> Feel free to add this or a screenshot where you can see a bit more of the page to the readme

I added a  image to the readme. Let me know if you like this one or one without a page header! (Essentially you won't have the placeholder BMX Racing event header)

pc-coholic commented:
Just a quick note since you added Apple Pay and Google Pay: In the past, we had a lot of users asking to please also display Apple Pay and Google Pay as a payment method on the payment selection step.

For that, we introduced the `WalletQueries`. They are by no means perfect, but do some rudimentary checks if a browser could - in theory - do Apple Pay or Google Pay payments.

See [this](https://github.com/pretix/pretix/blob/a90b40035c6c627f39e1d83bf248b7ada81f23dc/src/pretix/plugins/stripe/payment.py#L1108-L1114) for an example with no logic whatsoever - but if so desired, you could even check with SumUp first if the merchant is allowed to do Apple Pay/Google Pay and include that in the WalletQueries-result.

We also always add an option to disable this detection magic, since - of course - while a lot of people want it, there is also always some who don't ;-)
